### DISCUSSION TITLE
Support for dual mode aka OTG

### DISCUSSION THREAD

**Discussion by cjhilder** [hathach/tinyusb#1504] [Q&A] [answered]
I have built a board that connects to a laptop using USB and exchanges data via CDC. The board includes a TUSB320 chip, a VBUS switch, and a type-C USB receptacle, so I have all the hardware available to detect whether to go into host or device mode, and in host mode to turn on VBUS whilst keeping it off in device mode. Basically the TUSB320 will tell me whether tinyUSB should be in device or host mode.

My problem is that when I initialize tinyUSB I can't see any way to tell it whether to initialize as device or host. I want to have both device and host stacks linked into my code and to initialize one or the other.

There are no tinyUSB examples showing this. My question is, can it be done or is it beyond the scope of what tinyUSB can do?

**Comment by hathach** [comment, marked as answer]
it is possible but just not implemented yet.

**Reply by cjhilder** [reply]
Is it something that is on the horizon? My previous attempts at customizing the code base ended in tears so I am dependent on others to implement such things. Do you, or anyone else, have plans to implement this in the near future?

**Reply by hathach** [reply]
No, I have too many works now that I don't plan any more.

**Reply by cjhilder** [reply]
Thanks for your reply. Your work is much appreciated and I will find a way - possibly compiling two versions of my code, saving both in flash, and reboot from one to another as required to switch from device to host. It's a future project for me one day soon.

**Comment by ikskuh** [comment]
Heya!

Sad to hear that this is not supported yet, what work would be required to be done in order to be able to switch between host and device mode?

I have a project that can actively switch between two USB ports using the same PHY, but i can't reset the system as other operations are still in-flight. I'm willing to implement that as a PR or fork, but i need some hints on what to do (rough order of operations should be enough)

**Reply by ikskuh** [reply]
cc @hathach

**Reply by hathach** [reply]
it is rather simple and straight-forward, you only need to additional `rhport` arguments to all the APIs of tinyusb, from user -> class driver -> usbd to let dcd know which controller it should talk to. DCD API already has rhport as part of its argument already. Along the way, you would also need to add struct/variable to hold internal state for each controller. It is quite lots of work though.

**Reply by ikskuh** [reply]
That doesn't sound like it would solve the problem, as i still cannot call `tud_init(0)` after a `tuh_init(0)` was invoked. I want to use the same host port/PHY for both device mode and host mode, switchable at runtime. This means i need something like `tuh_deinit()` and `tud_deinit()` which will revoke the previous initialization and allows reinitialization at a later point. This is also useful for power saving.

**Reply by hathach** [reply]
oh you are right, sorry, I mixed up with other issue for multiple devices on multiple ports. For switching mode, right, you will need to configure the controller from host <-> device and vice versa e.g depending on the ID pin state. It should be a lighter version of deinit() since we only want to clear the state and set controller mode rather than destroy the whole stack. 

For hints: you will need to go through codes and make changes as needed (no specific hint I know). You should look at both tud_init()/dcd_init() and tuh_init()/hcd_init() and/or add an tusb_set_role()/otg_set_role() for stack/controller role switch if needed. Each port will be a bit different though.

**Reply by ikskuh** [reply]
Thank you!

I'll try implementing it. If i get it working stable, i'll make a PR, happy for review then :)

**Reply by DRNadler** [reply]
@MasterQ32 and @cjhilder - Were you able to get OTG working? I am starting a project on a board with OTG implemented, and I need to switch between device-CDC and host-MSC. Thanks!

**Reply by ikskuh** [reply]
Yes, i got OTG for the rp2040 working, see #1835.

It's still a bit rough on the edges, but you can init and deinit both host and client device in that PR. you probably need to implement the deinit funcs for the rhport tho

**Reply by DRNadler** [reply]
@MasterQ32 Thanks, I will have a look...

**Reply by DRNadler** [reply]
@MasterQ32 - I just cloned the tinyUSB repo and could not apply your patch; sorry I'm not a GIT expert and don't know what this means, any idea??

c:\evkT1\tinyusb>git am "Z:\Downloads\vario\1835.patch"
C:/evkT1/.git/modules/tinyusb/rebase-apply/patch:262: trailing whitespace.
 
error: patch failed: src/device/usbd.c:379
error: src/device/usbd.c: patch does not apply
error: patch failed: src/host/usbh.c:348
error: src/host/usbh.c: patch does not apply
Applying: Implements tuh_deinit() and tud_deinit() to uninitialize host/device mode.
Patch failed at 0001 Implements tuh_deinit() and tud_deinit() to uninitialize host/device mode.
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
hint: Use 'git am --show-current-patch=diff' to see the failed patch

c:\evkT1\tinyusb>
